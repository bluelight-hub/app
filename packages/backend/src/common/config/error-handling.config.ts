import { DuplicateDetectionConfig } from '../utils/duplicate-detection.util';
import { RetryConfig } from '../utils/retry.util';

/**
 * Umgebungsspezifische Konfiguration für Error Handling
 */
export interface EnvironmentErrorConfig {
    /** Retry-Konfiguration für diese Umgebung */
    retryConfig: RetryConfig;
    /** Duplicate Detection Konfiguration */
    duplicateDetectionConfig: DuplicateDetectionConfig;
    /** Ob detaillierte Fehlerberichte aktiviert sind */
    verboseErrorReporting: boolean;
    /** Ob strikte Validierung aktiviert ist */
    strictValidation: boolean;
    /** Ob Feature Flags aktiviert sind */
    enableFeatureFlags: boolean;
    /** Maximale Anzahl gleichzeitiger Operationen */
    maxConcurrentOperations: number;
    /** Timeout für Operationen in Millisekunden */
    operationTimeout: number;
    /** Ob Fehler-Metriken gesammelt werden sollen */
    collectErrorMetrics: boolean;
}

/**
 * Development Environment Konfiguration
 */
export const DEVELOPMENT_ERROR_CONFIG: EnvironmentErrorConfig = {
    retryConfig: {
        maxRetries: 5, // Mehr Versuche für Entwicklung
        baseDelay: 1000,
        maxDelay: 15000,
        backoffMultiplier: 2,
        jitterFactor: 0.2,
        timeout: 60000, // Längere Timeouts für Debugging
    },
    duplicateDetectionConfig: {
        timeWindow: 120000, // 2 Minuten für Entwicklung
        maxCacheSize: 500,
        cleanupInterval: 60000, // 1 Minute
    },
    verboseErrorReporting: true,
    strictValidation: false, // Weniger strikt für Entwicklung
    enableFeatureFlags: true,
    maxConcurrentOperations: 10,
    operationTimeout: 60000,
    collectErrorMetrics: true,
};

/**
 * Test Environment Konfiguration
 */
export const TEST_ERROR_CONFIG: EnvironmentErrorConfig = {
    retryConfig: {
        maxRetries: 2, // Weniger Versuche für schnelle Tests
        baseDelay: 100,
        maxDelay: 1000,
        backoffMultiplier: 2,
        jitterFactor: 0.1,
        timeout: 5000, // Kurze Timeouts für Tests
    },
    duplicateDetectionConfig: {
        timeWindow: 10000, // 10 Sekunden für Tests
        maxCacheSize: 100,
        cleanupInterval: 5000, // 5 Sekunden
    },
    verboseErrorReporting: false,
    strictValidation: true, // Strikt für Tests
    enableFeatureFlags: false,
    maxConcurrentOperations: 5,
    operationTimeout: 5000,
    collectErrorMetrics: false,
};

/**
 * Production Environment Konfiguration
 */
export const PRODUCTION_ERROR_CONFIG: EnvironmentErrorConfig = {
    retryConfig: {
        maxRetries: 3, // Moderate Anzahl für Production
        baseDelay: 500,
        maxDelay: 10000,
        backoffMultiplier: 2,
        jitterFactor: 0.15,
        timeout: 30000, // Standard Timeouts
    },
    duplicateDetectionConfig: {
        timeWindow: 60000, // 1 Minute für Production
        maxCacheSize: 1000,
        cleanupInterval: 300000, // 5 Minuten
    },
    verboseErrorReporting: false, // Keine detaillierten Fehler in Production
    strictValidation: true,
    enableFeatureFlags: true,
    maxConcurrentOperations: 20,
    operationTimeout: 30000,
    collectErrorMetrics: true,
};

/**
 * Staging Environment Konfiguration
 */
export const STAGING_ERROR_CONFIG: EnvironmentErrorConfig = {
    retryConfig: {
        maxRetries: 3,
        baseDelay: 750,
        maxDelay: 12000,
        backoffMultiplier: 2,
        jitterFactor: 0.15,
        timeout: 45000,
    },
    duplicateDetectionConfig: {
        timeWindow: 90000, // 1.5 Minuten für Staging
        maxCacheSize: 750,
        cleanupInterval: 180000, // 3 Minuten
    },
    verboseErrorReporting: true, // Detaillierte Fehler für Staging
    strictValidation: true,
    enableFeatureFlags: true,
    maxConcurrentOperations: 15,
    operationTimeout: 45000,
    collectErrorMetrics: true,
};

/**
 * Gibt die Konfiguration für die aktuelle Umgebung zurück
 * 
 * @param environment Die Umgebung (NODE_ENV)
 * @returns Die entsprechende Error Handling Konfiguration
 */
export function getErrorConfigForEnvironment(environment: string): EnvironmentErrorConfig {
    switch (environment?.toLowerCase()) {
        case 'development':
        case 'dev':
            return DEVELOPMENT_ERROR_CONFIG;
        case 'test':
        case 'testing':
            return TEST_ERROR_CONFIG;
        case 'production':
        case 'prod':
            return PRODUCTION_ERROR_CONFIG;
        case 'staging':
        case 'stage':
            return STAGING_ERROR_CONFIG;
        default:
            // Fallback auf Development für unbekannte Umgebungen
            return DEVELOPMENT_ERROR_CONFIG;
    }
}

/**
 * Feature Flags für Error Handling
 */
export interface ErrorHandlingFeatureFlags {
    /** Ob erweiterte Retry-Logik aktiviert ist */
    enableAdvancedRetry: boolean;
    /** Ob Duplicate Detection aktiviert ist */
    enableDuplicateDetection: boolean;
    /** Ob Metriken-Sammlung aktiviert ist */
    enableMetricsCollection: boolean;
    /** Ob Circuit Breaker aktiviert ist */
    enableCircuitBreaker: boolean;
    /** Ob Rate Limiting aktiviert ist */
    enableRateLimiting: boolean;
}

/**
 * Standard Feature Flags
 */
export const DEFAULT_FEATURE_FLAGS: ErrorHandlingFeatureFlags = {
    enableAdvancedRetry: true,
    enableDuplicateDetection: true,
    enableMetricsCollection: true,
    enableCircuitBreaker: false, // Noch nicht implementiert
    enableRateLimiting: false, // Noch nicht implementiert
}; 