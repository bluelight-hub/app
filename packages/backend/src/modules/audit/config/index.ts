export * from './audit-interceptor.config';
