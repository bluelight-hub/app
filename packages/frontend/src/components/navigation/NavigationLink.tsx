import React from 'react';
import { Link, LinkProps } from 'react-router-dom';
import { Button, Tooltip } from 'antd';
import { useActiveNavigation } from '@/hooks/useActiveNavigation';
import { useAuthorization } from '@/hooks/useAuthorization';

export interface NavigationLinkProps extends Omit<LinkProps, 'to'> {
  /** Ziel-URL */
  to: string;
  /** Anzuzeigender Text */
  children: React.ReactNode;
  /** Icon für den Link */
  icon?: React.ReactNode;
  /** Zusätzliche CSS-Klassen für aktive Links */
  activeClassName?: string;
  /** Soll exakte Übereinstimmung geprüft werden? */
  exact?: boolean;
  /** Ist der Link deaktiviert? */
  disabled?: boolean;
  /** Tooltip-Text */
  tooltip?: string;
  /** Button-Stil verwenden */
  asButton?: boolean;
  /** Button-Typ (falls asButton = true) */
  buttonType?: 'default' | 'primary' | 'ghost' | 'dashed' | 'link' | 'text';
  /** Button-Größe (falls asButton = true) */
  buttonSize?: 'small' | 'middle' | 'large';
  /** Callback für Klick-Events */
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
  /** Prüfe Zugriffsberechtigung automatisch */
  checkAccess?: boolean;
  /** Fallback-Komponente bei fehlender Berechtigung */
  unauthorizedFallback?: React.ReactNode;
}

/**
 * Erweiterte NavigationLink-Komponente mit automatischer Active-State-Erkennung
 * und Zugriffsberechtigungsprüfung
 */
const NavigationLink: React.FC<NavigationLinkProps> = ({
  to,
  children,
  icon,
  className = '',
  activeClassName = 'active',
  exact = false,
  disabled = false,
  tooltip,
  asButton = false,
  buttonType = 'default',
  buttonSize = 'middle',
  onClick,
  checkAccess = true,
  unauthorizedFallback = null,
  ...linkProps
}) => {
  const { isExactActive, isActiveOrChild, getLinkClasses } = useActiveNavigation();
  const { canAccessPath } = useAuthorization();

  // Prüfe Aktivitätsstatus
  const isActive = exact ? isExactActive(to) : isActiveOrChild(to);

  // Prüfe Zugriffsberechtigung
  const hasAccess = !checkAccess || canAccessPath(to);

  // Handle Click
  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    if (disabled || !hasAccess) {
      event.preventDefault();
      return;
    }

    if (onClick) {
      onClick(event);
    }
  };

  // Zeige Fallback bei fehlender Berechtigung
  if (!hasAccess && unauthorizedFallback) {
    return <>{unauthorizedFallback}</>;
  }

  // Generiere CSS-Klassen
  const linkClasses = getLinkClasses(to, `navigation-link ${className}`, `navigation-link--active ${activeClassName}`);

  // Zusätzliche Klassen für disabled/unauthorized
  const finalClasses = `
    ${linkClasses}
    ${disabled ? 'navigation-link--disabled' : ''}
    ${!hasAccess ? 'navigation-link--unauthorized' : ''}
    ${isActive ? 'navigation-link--current' : ''}
  `.trim();

  // Button-Stil
  if (asButton) {
    const buttonElement = (
      <Button
        type={isActive ? 'primary' : buttonType}
        size={buttonSize}
        icon={icon}
        disabled={disabled || !hasAccess}
        className={finalClasses}
        onClick={handleClick}
      >
        {children}
      </Button>
    );

    // Wrapper um Button für Navigation
    const buttonWithNavigation = (
      <Link
        to={to}
        {...linkProps}
        className="navigation-link-wrapper"
        onClick={handleClick}
        style={{ textDecoration: 'none' }}
      >
        {buttonElement}
      </Link>
    );

    return tooltip ? (
      <Tooltip title={tooltip} placement="top">
        {buttonWithNavigation}
      </Tooltip>
    ) : (
      buttonWithNavigation
    );
  }

  // Standard Link-Stil
  const linkElement = (
    <Link
      to={to}
      {...linkProps}
      className={finalClasses}
      onClick={handleClick}
      aria-current={isActive ? 'page' : undefined}
      aria-disabled={disabled || !hasAccess}
    >
      {icon && <span className="navigation-link__icon">{icon}</span>}
      <span className="navigation-link__text">{children}</span>
    </Link>
  );

  return tooltip ? (
    <Tooltip title={tooltip} placement="right">
      {linkElement}
    </Tooltip>
  ) : (
    linkElement
  );
};

export default NavigationLink;
