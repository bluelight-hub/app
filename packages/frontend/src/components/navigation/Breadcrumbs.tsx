import React, { useMemo } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { Breadcrumb } from 'antd';
import { HomeOutlined } from '@ant-design/icons';
import { RouteUtils } from '@/config/routes';
import { useAuthorization } from '@/hooks/useAuthorization';

export interface BreadcrumbItem {
  /** Label für den Breadcrumb */
  label: string;
  /** Pfad für die Navigation */
  path: string;
  /** Ist der Breadcrumb klickbar? */
  clickable: boolean;
  /** Icon für den Breadcrumb */
  icon?: React.ReactNode;
  /** Ist dies der aktuelle/letzte Breadcrumb? */
  current?: boolean;
}

export interface BreadcrumbsProps {
  /** Zusätzliche CSS-Klassen */
  className?: string;
  /** Custom Separator zwischen Breadcrumbs */
  separator?: React.ReactNode;
  /** Zeige Home-Icon beim ersten Element */
  showHomeIcon?: boolean;
  /** Maximum Anzahl der angezeigten Breadcrumbs */
  maxItems?: number;
  /** Custom Breadcrumb-Items (überschreibt automatische Generierung) */
  customItems?: BreadcrumbItem[];
  /** Callback wenn ein Breadcrumb geklickt wird */
  onItemClick?: (item: BreadcrumbItem) => void;
}

/**
 * Dynamische Breadcrumb-Komponente
 * Generiert automatisch Breadcrumbs basierend auf der aktuellen Route und Route-Metadaten
 */
const Breadcrumbs: React.FC<BreadcrumbsProps> = ({
  className,
  separator = '/',
  showHomeIcon = true,
  maxItems = 5,
  customItems,
  onItemClick,
}) => {
  const location = useLocation();
  const { canAccessPath } = useAuthorization();

  // Generiere Breadcrumb-Items basierend auf der aktuellen Route
  const breadcrumbItems = useMemo((): BreadcrumbItem[] => {
    // Verwende custom Items falls vorhanden
    if (customItems) {
      return customItems;
    }

    // Generiere automatisch basierend auf der Route
    const items = RouteUtils.generateBreadcrumbs(location.pathname);

    return items.map((item, index) => ({
      ...item,
      current: index === items.length - 1,
      icon: index === 0 && showHomeIcon ? <HomeOutlined /> : undefined,
    }));
  }, [location.pathname, customItems, showHomeIcon]);

  // Begrenze die Anzahl der angezeigten Items
  const displayItems = useMemo(() => {
    if (breadcrumbItems.length <= maxItems) {
      return breadcrumbItems;
    }

    // Zeige immer das erste und die letzten Items
    const start = breadcrumbItems.slice(0, 1);
    const end = breadcrumbItems.slice(-(maxItems - 2));

    return [
      ...start,
      {
        label: '...',
        path: '',
        clickable: false,
        current: false,
      },
      ...end,
    ];
  }, [breadcrumbItems, maxItems]);

  // Rendere einzelnen Breadcrumb-Item
  const renderBreadcrumbItem = (item: BreadcrumbItem, _index: number) => {
    const handleClick = () => {
      if (onItemClick) {
        onItemClick(item);
      }
    };

    // Nicht klickbare Items oder Ellipsis
    if (!item.clickable || item.label === '...' || item.current) {
      return (
        <span
          className={`${item.current ? 'font-medium text-gray-900' : 'text-gray-500'}`}
          onClick={item.clickable ? handleClick : undefined}
        >
          {item.icon && <span className="mr-1">{item.icon}</span>}
          {item.label}
        </span>
      );
    }

    // Prüfe Zugriffsberechtigung für den Link
    const canAccess = canAccessPath(item.path);

    if (!canAccess) {
      return (
        <span className="text-gray-400 cursor-not-allowed">
          {item.icon && <span className="mr-1">{item.icon}</span>}
          {item.label}
        </span>
      );
    }

    // Klickbarer Link
    return (
      <Link
        to={item.path}
        className="text-blue-600 hover:text-blue-800 transition-colors duration-200"
        onClick={handleClick}
      >
        {item.icon && <span className="mr-1">{item.icon}</span>}
        {item.label}
      </Link>
    );
  };

  // Konvertiere zu Ant Design Breadcrumb Items
  const antdItems = displayItems.map((item, index) => ({
    title: renderBreadcrumbItem(item, index),
    key: item.path || `item-${index}`,
  }));

  // Zeige keine Breadcrumbs wenn keine Items vorhanden
  if (displayItems.length === 0) {
    return null;
  }

  return (
    <div className={`breadcrumbs-container ${className || ''}`}>
      <Breadcrumb separator={separator} items={antdItems} className="text-sm" />
    </div>
  );
};

export default Breadcrumbs;
