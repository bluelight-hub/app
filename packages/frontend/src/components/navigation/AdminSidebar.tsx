import React, { useMemo } from 'react';
import { Avatar, Dropdown, Layout, Menu, Typography } from 'antd';
import { LogoutOutlined, MenuFoldOutlined, MenuUnfoldOutlined, SettingOutlined, UserOutlined } from '@ant-design/icons';
import { useNavigate } from 'react-router-dom';
import { RouteDefinition, RouteUtils } from '@/config/routes';
import { useAuthorization } from '@/hooks/useAuthorization';
import { useActiveNavigation } from '@/hooks/useActiveNavigation';
import NavigationLink from './NavigationLink';
import { useAuth } from '@/hooks/useAuth';
import { MenuItem } from '@/utils/navigationConverter.ts';

const { Sider } = Layout;
const { Text } = Typography;

export interface AdminSidebarProps {
  /** Ist die Sidebar zusammengeklappt? */
  collapsed?: boolean;
  /** Callback für Collapse-Status-Änderung */
  onCollapse?: (collapsed: boolean) => void;
  /** Zusätzliche CSS-Klassen */
  className?: string;
  /** Breite der Sidebar */
  width?: number;
  /** Breite der zusammengeklappten Sidebar */
  collapsedWidth?: number;
}

/**
 * Admin Navigation Sidebar mit rollenbasierter Menüfilterung
 */
const AdminSidebar: React.FC<AdminSidebarProps> = ({
  collapsed = false,
  onCollapse,
  className,
  width = 256,
  collapsedWidth = 80,
}) => {
  const navigate = useNavigate();
  const { user, logout } = useAuth();
  const { authorizedRoutes } = useAuthorization();
  const { currentPath } = useActiveNavigation();

  // Filtere und gruppiere autorisierte Routen für die Navigation
  const navigationItems = useMemo(() => {
    const routes = authorizedRoutes.filter((route) => route.showInNavigation);
    const grouped = RouteUtils.getRoutesByCategory();

    return Object.entries(grouped)
      .map(([category, categoryRoutes]) => {
        const authorizedCategoryRoutes = categoryRoutes.filter((route) =>
          routes.some((authRoute) => authRoute.id === route.id),
        );

        if (authorizedCategoryRoutes.length === 0) return null;

        return {
          category,
          routes: authorizedCategoryRoutes,
        };
      })
      .filter(Boolean);
  }, [authorizedRoutes]);

  // Konvertiere Routen zu Ant Design Menu Items
  const menuItems = useMemo(() => {
    const items: MenuItem[] = [];

    navigationItems.forEach((group, groupIndex) => {
      if (!group) return;

      // Füge Kategorie-Header hinzu (nur wenn nicht collapsed)
      if (!collapsed && group.routes.length > 0) {
        items.push({
          type: 'group',
          label: getCategoryDisplayName(group.category),
          key: `group-${group.category}`,
        });
      }

      // Füge Routen hinzu
      group.routes.forEach((route: RouteDefinition) => {
        items.push({
          key: route.path,
          icon: getRouteIcon(route.icon),
          label: (
            <NavigationLink
              to={route.path}
              checkAccess={false} // Bereits in authorizedRoutes gefiltert
              className="w-full"
            >
              {route.title}
            </NavigationLink>
          ),
          title: route.description,
        });
      });

      // Füge Divider zwischen Gruppen hinzu (aber nicht nach der letzten)
      if (groupIndex < navigationItems.length - 1 && !collapsed) {
        items.push({
          type: 'divider',
          key: `divider-${groupIndex}`,
        });
      }
    });

    return items;
  }, [navigationItems, collapsed]);

  // User Dropdown Menu
  const userMenuItems = [
    {
      key: 'profile',
      icon: <UserOutlined />,
      label: 'Profil',
      onClick: () => navigate('/admin/profile'),
    },
    {
      key: 'settings',
      icon: <SettingOutlined />,
      label: 'Einstellungen',
      onClick: () => navigate('/admin/settings'),
    },
    {
      type: 'divider' as const,
    },
    {
      key: 'logout',
      icon: <LogoutOutlined />,
      label: 'Abmelden',
      onClick: logout,
    },
  ];

  // Bestimme aktuell ausgewählte Menu Items
  const selectedKeys = useMemo(() => {
    const route = RouteUtils.findByPath(currentPath);
    return route ? [route.path] : [];
  }, [currentPath]);

  return (
    <Sider
      width={width}
      collapsedWidth={collapsedWidth}
      collapsed={collapsed}
      onCollapse={onCollapse}
      className={`admin-sidebar ${className || ''}`}
      style={{
        overflow: 'auto',
        height: '100vh',
        position: 'fixed',
        left: 0,
        top: 0,
        bottom: 0,
        zIndex: 1000,
      }}
    >
      {/* Header */}
      <div className="admin-sidebar__header p-4 border-b border-gray-200">
        <div className="flex items-center justify-between">
          {!collapsed && (
            <div className="flex items-center space-x-2">
              <div className="w-8 h-8 bg-blue-600 rounded-lg flex items-center justify-center">
                <span className="text-white font-bold text-sm">BH</span>
              </div>
              <div>
                <Text className="font-semibold text-white">Bluelight Hub</Text>
                <br />
                <Text className="text-xs text-gray-300">Admin Panel</Text>
              </div>
            </div>
          )}

          <button onClick={() => onCollapse?.(!collapsed)} className="p-1 rounded hover:bg-gray-700 text-white">
            {collapsed ? <MenuUnfoldOutlined /> : <MenuFoldOutlined />}
          </button>
        </div>
      </div>

      {/* Navigation Menu */}
      <div className="admin-sidebar__navigation flex-1">
        <Menu
          mode="inline"
          theme="dark"
          selectedKeys={selectedKeys}
          items={menuItems}
          inlineCollapsed={collapsed}
          className="border-none"
        />
      </div>

      {/* User Profile Footer */}
      <div className="admin-sidebar__footer p-4 border-t border-gray-700">
        {collapsed ? (
          <Dropdown menu={{ items: userMenuItems }} placement="topRight" trigger={['click']}>
            <Avatar icon={<UserOutlined />} className="cursor-pointer w-full flex items-center justify-center" />
          </Dropdown>
        ) : (
          <Dropdown menu={{ items: userMenuItems }} placement="topRight" trigger={['click']}>
            <div className="flex items-center space-x-3 cursor-pointer hover:bg-gray-700 p-2 rounded">
              <Avatar icon={<UserOutlined />} />
              <div className="flex-1 min-w-0">
                <Text className="text-white font-medium block truncate">{user?.name || 'Admin User'}</Text>
                <Text className="text-gray-300 text-xs block truncate">{user?.role || 'Administrator'}</Text>
              </div>
            </div>
          </Dropdown>
        )}
      </div>
    </Sider>
  );
};

// Helper Funktionen
function getCategoryDisplayName(category: string): string {
  const categoryNames: Record<string, string> = {
    overview: 'Übersicht',
    management: 'Verwaltung',
    system: 'System',
    monitoring: 'Überwachung',
    tools: 'Tools',
    analytics: 'Analytics',
    configuration: 'Konfiguration',
    other: 'Weitere',
  };

  return categoryNames[category] || category;
}

function getRouteIcon(iconName?: string): React.ReactNode {
  // Diese Funktion würde die entsprechenden Icons basierend auf dem Namen zurückgeben
  // Implementierung abhängig von der verwendeten Icon-Bibliothek

  const iconMap: Record<string, React.ReactNode> = {
    House: <UserOutlined />, // Placeholder - echte Icons verwenden
    Users: <UserOutlined />,
    Buildings: <UserOutlined />,
    Gear: <SettingOutlined />,
    FileText: <UserOutlined />,
    Database: <UserOutlined />,
    Code: <UserOutlined />,
    ChartBar: <UserOutlined />,
    Sliders: <SettingOutlined />,
  };

  return iconName ? iconMap[iconName] || <UserOutlined /> : <UserOutlined />;
}

export default AdminSidebar;
