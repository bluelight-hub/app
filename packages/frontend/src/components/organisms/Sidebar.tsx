import { Button, Menu, MenuProps } from 'antd';
import React from 'react';
import { PiTextOutdent } from 'react-icons/pi';
import { useLocation, useNavigate } from 'react-router';
import { NavigationItem, workspaces } from '../../config/navigation';
import { useThemeInternal } from '../../hooks/useTheme';
import Divider from '../atoms/Divider';
import Logo from '../atoms/Logo';
import UserProfile from '../atoms/UserProfile';

interface SidebarProps {
    isOpen?: boolean;
    onClose?: () => void;
    isMobile?: boolean;
    navigation: NavigationItem[];
}

type MenuItem = Required<MenuProps>['items'][number];

const convertNavigationToMenuItems = (items: NavigationItem[]): MenuItem[] => {
    return items.map((item) => {
        if (item.type === 'item') {
            return {
                key: item.key,
                icon: item.icon && React.createElement(item.icon, { size: 20 }),
                label: item.label,
            } as MenuItem;
        }

        if (item.type === 'group') {
            return {
                type: 'group',
                key: item.key,
                label: item.label,
                children: item.children.map((child) => {
                    if (child.type === 'item') {
                        return {
                            key: child.key,
                            icon: child.icon && React.createElement(child.icon, { size: 20 }),
                            label: child.label,
                        } as MenuItem;
                    }
                    if (child.type === 'submenu') {
                        return {
                            type: 'submenu',
                            key: child.key,
                            icon: child.icon && React.createElement(child.icon, { size: 20 }),
                            label: child.label,
                            children: child.children.map((subChild) => ({
                                key: subChild.key,
                                icon: subChild.icon && React.createElement(subChild.icon, { size: 20 }),
                                label: subChild.label,
                            } as MenuItem)),
                        } as MenuItem;
                    }
                    return null;
                }).filter(Boolean) as MenuItem[],
            } as MenuItem;
        }

        if (item.type === 'submenu') {
            return {
                type: 'submenu',
                key: item.key,
                icon: item.icon && React.createElement(item.icon, { size: 20 }),
                label: item.label,
                children: item.children.map((child) => ({
                    key: child.key,
                    icon: child.icon && React.createElement(child.icon, { size: 20 }),
                    label: child.label,
                } as MenuItem)),
            } as MenuItem;
        }

        return null;
    }).filter(Boolean) as MenuItem[];
};

/**
 * Main sidebar component containing navigation and user profile
 */
const Sidebar: React.FC<SidebarProps> = ({ isOpen, onClose, isMobile = false, navigation }) => {
    const navigate = useNavigate();
    const location = useLocation();

    const handleMenuClick: MenuProps['onClick'] = (info) => {
        navigate(info.key);
        if (isMobile && onClose) {
            onClose();
        }
    };

    // Pr√ºfen ob Dark Mode aktiv ist
    const themeUtils = useThemeInternal();

    const sidebarContent = (
        <div className="flex h-screen flex-col gap-y-5 border-r border-gray-200 dark:border-gray-800 bg-white dark:bg-gray-900">
            <div className="flex h-16 shrink-0 items-center gap-x-3 px-6">
                <Logo className="h-8 w-auto" />
                <span className="text-xl font-semibold text-gray-900 dark:text-white">Bluelight Hub</span>
            </div>
            <Divider className="-mt-3 mb-3" />
            <nav className="flex flex-1 flex-col">
                <div className="h-[calc(100vh-9.5rem)] overflow-y-auto">
                    <Menu
                        theme={themeUtils.isDark ? 'dark' : 'light'}
                        mode="inline"
                        selectedKeys={[location.pathname]}
                        onClick={handleMenuClick}
                        items={convertNavigationToMenuItems(navigation)}
                        className="border-none bg-transparent"
                    />
                    <Divider />
                    <Menu
                        theme={themeUtils.isDark ? 'dark' : 'light'}
                        mode="inline"
                        selectedKeys={[location.pathname]}
                        onClick={handleMenuClick}
                        items={convertNavigationToMenuItems(workspaces)}
                        className="border-none bg-transparent"
                    />
                </div>
                <div className="mt-auto w-full shrink-0 overflow-hidden">
                    <UserProfile href="#" />
                </div>
            </nav>
        </div>
    );

    if (isMobile && isOpen) {
        return (
            <div className="fixed inset-0 z-50 lg:hidden" role="dialog" aria-modal="true">
                <>
                    {/* Backdrop */}
                    <div
                        className="fixed inset-0 bg-gray-900/80 dark:bg-black/80 cursor-pointer"
                        onClick={onClose}
                    />

                    {/* Sidebar Container */}
                    <div className="fixed inset-y-0 left-0 flex max-w-xs w-full">
                        {/* Main Sidebar */}
                        <div className="relative flex-1 flex w-full">
                            {sidebarContent}

                            {/* Close Button */}
                            <div className="absolute right-0 top-0 -mr-16 pt-4">
                                <Button
                                    type="text"
                                    className="text-white hover:text-gray-200 flex items-center justify-center rounded-md focus:outline-none focus:ring-2 focus:ring-white"
                                    onClick={onClose}
                                >
                                    <span className="sr-only">Close sidebar</span>
                                    <PiTextOutdent size={24} />
                                </Button>
                            </div>
                        </div>
                    </div>
                </>
            </div>
        );
    }

    return (
        <div className="hidden lg:fixed lg:inset-y-0 lg:z-50 lg:flex lg:w-72 lg:flex-col">
            {sidebarContent}
        </div>
    );
};

export default Sidebar; 