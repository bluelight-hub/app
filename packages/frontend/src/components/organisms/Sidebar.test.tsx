import { render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { NavigationItem } from '../../config/navigation';
import Sidebar from './Sidebar';

// Mock matchMedia
beforeEach(() => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    })),
  });
});

// Mock MainSidebar component
vi.mock('../molecules/sidebar/MainSidebar', () => ({
  __esModule: true,
  default: vi.fn(({ isOpen, onClose, isMobile, navigation }) => (
    <div data-testid="main-sidebar">
      <div data-testid="props-display">
        isOpen: {String(isOpen)}, isMobile: {String(isMobile)}, navItems: {navigation.length}
      </div>
      <button data-testid="close-button" onClick={onClose}>
        Close
      </button>
    </div>
  )),
}));

const mockNavigation: NavigationItem[] = [
  {
    type: 'item',
    key: '/test',
    path: '/test',
    label: 'Test Item',
  },
];

describe('Sidebar', () => {
  it('should render MainSidebar with all props', () => {
    const mockOnClose = vi.fn();

    render(<Sidebar isOpen={true} onClose={mockOnClose} isMobile={true} navigation={mockNavigation} />);

    // Check if MainSidebar is rendered
    expect(screen.getByTestId('main-sidebar')).toBeInTheDocument();

    // Check if props were passed correctly
    expect(screen.getByTestId('props-display')).toHaveTextContent('isOpen: true');
    expect(screen.getByTestId('props-display')).toHaveTextContent('isMobile: true');
    expect(screen.getByTestId('props-display')).toHaveTextContent('navItems: 1');
  });

  it('should pass onClose function to MainSidebar', () => {
    const mockOnClose = vi.fn();

    render(<Sidebar isOpen={true} onClose={mockOnClose} navigation={mockNavigation} />);

    // Trigger close function
    screen.getByTestId('close-button').click();

    // Verify onClose was called
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  // Snapshot test
  it('should match snapshot', () => {
    const { container } = render(<Sidebar navigation={mockNavigation} />);

    expect(container).toMatchSnapshot();
  });
});
