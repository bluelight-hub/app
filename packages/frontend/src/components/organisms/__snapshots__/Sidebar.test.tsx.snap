// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > should match snapshot 1`] = `
<div>
  <div
    data-testid="main-sidebar"
  >
    <div
      data-testid="props-display"
    >
      isOpen: 
      undefined
      , isMobile: 
      undefined
      , navItems: 
      1
    </div>
    <button
      data-testid="close-button"
    >
      Close
    </button>
  </div>
</div>
`;
